Framework/Kernel/inc/MantidKernel/Strings.h

```diff
@@ -75,6 +79,113 @@ DLLExport std::string join(ITERATOR_TYPE begin, ITERATOR_TYPE end,
+ *  out = join(v.begin(), v.end(), ", ");
+ *
+ * This version is used for random access iterators (e.g. map, set), and
+ * it calls simpleJoin().
+ *
+ * @param begin :: iterator at the start
+ * @param end :: iterator at the end
+ * @param separator :: string to append.
+ * @return
+ */
+template <typename ITERATOR_TYPE>
+DLLExport std::string
+join(ITERATOR_TYPE begin, ITERATOR_TYPE end, const std::string &separator,
+     typename std::enable_if<
+         !(std::is_same<
+             typename std::iterator_traits<ITERATOR_TYPE>::iterator_category,
+             std::random_access_iterator_tag>::value)>::type * = nullptr) {
+  return simpleJoin(begin, end, separator);
+}
+
+//------------------------------------------------------------------------------------------------
+/** Join a set or vector of (something that turns into a string) together
+ * into one string, separated by a string.
+ * Returns an empty string if the range is null.
+ * Does not add the separator after the LAST item.
+ *
+ * For example, join a vector of strings with commas with:
+ *  out = join(v.begin(), v.end(), ", ");
+ *
+ * This is a faster threaded version of the join() function above.
+ * It is used only if the iterators are not random access (e.g. vector), as it
+ * needs to be able to determine the distance between begin and end.
+ * It reverts to calling simpleJoin() if the input array is small.
+ *
+ * @param begin :: iterator at the start
+ * @param end :: iterator at the end
+ * @param separator :: string to append.
+ * @return
+ */
+template <typename ITERATOR_TYPE>
+DLLExport std::string
+join(ITERATOR_TYPE begin, ITERATOR_TYPE end, const std::string &separator,
+     typename std::enable_if<
+         (std::is_same<
+             typename std::iterator_traits<ITERATOR_TYPE>::iterator_category,
+             std::random_access_iterator_tag>::value)>::type * = nullptr) {
+
+  // Get max number of threads
+  int nmaxThreads = static_cast<int>(PARALLEL_GET_MAX_THREADS);
+
+  // Define minimum size for using threading
+  int min_size = 500 * nmaxThreads;
+
+  // Get the distance between begining and end
+  int dist = static_cast<int>(std::distance(begin, end));
+
+  if (dist < min_size) {
+
+    // If the input array is small, use the simpler function to avoid
+    // unnecessary overhead from generating the parallel section
+    return simpleJoin(begin, end, separator);
+
+  } else {
+
+    // Allocate vector space
+    std::vector<std::string> output(nmaxThreads);
+    size_t stream_size = 0;
+
+    // Actual number of threads in the current region
+    int nThreads = 1;
+#pragma omp parallel reduction(+ : stream_size)
+    {
+      nThreads = static_cast<int>(PARALLEL_NUMBER_OF_THREADS);
+      int idThread = static_cast<int>(PARALLEL_THREAD_NUMBER);
+      ITERATOR_TYPE it;
+
+      // Initialise ostringstream
+      std::ostringstream thread_stream;
+#pragma omp for schedule(static)
+      for (int i = 0; i < dist; i++) {
```

Comment:
I'm concerned that this implementation no longer preserves the order of elements in the container. Changing this may cause subtle runtime issues and unreliable tests.

Reply:
@nvaytet How about a `#pragma omp parallel {}`, inside that compute indices by hand (`size_t start = size/thread_count;`, ...), then the loop doing the work also within that block?

Reply:
@SimonHeybrock Yes I thought about doing this, but then @igudich pointed out that that should be exactly the same as what a static schedule does.
From the OpenMP specifications document:
"When schedule(static, chunk_size) is specified, iterations are divided into chunks of size chunk_size, and the chunks are assigned to the threads in the team in a round-robin fashion in the order of the thread number."
So it does say "in the order of the thread number".
In the pdf document (https://www.openmp.org/wp-content/uploads/openmp-4.5.pdf ; page 60), the argument "chunk_size" is in italics (and is an optional argument), so I am guessing the same ordering is applied even if chunk_size is not specified (this was not super clear just from the sentence above). If it is not specified, the pragma tries to divide the work into pieces of the same size.

Reply:
"When no chunk_size is specified, the iteration space is divided into chunks
that are approximately equal in size, and at most one chunk is distributed to
each thread. The size of the chunks is unspecified in this case"